Expose additional beaker caching backends

claudiofr opened the pull request:
Closes #15216

Convert beaker mulled resolution, citations and biotools service caches to use the database rather than the file system.

Introduce new config parameters for the beaker cache database url, schema name and cache table name.

Introduce logic to default values of a parameter to values of another parameter. i.e. default cache database url to database_connection. Metadata for this was placed in code. Should probably be placed eventually in config file which would require a change to the config_schema.yml.
Introduced a new unit tests, test_mulled_resolution_cache_db.py, test_citations_db.py
Did not introduce a new test for the biotools_service beaker cache because I could not find a place in the code where it was used.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

claudiofr committed:
Change muled resolution beaker cache type from file to ext:database.

Rather than store cache data in the file system store cache data
in the database. Partially addresses issue 15216. For now this was
only changed for the mulled resolution cache. Introduced new
config parameters for database url, cache table name. Introduced
logic to default values of a parameter to values of another
parameter. i.e. default cache database url to database_connection.
Metadata for this was placed in code. Should probably be
placed eventually in config file which would require a change
to the config_schema.yml.

Introduced a new unit test, test_mulled_resolution_cache_db.py.

claudiofr committed:
Change citation and biotools_service beaker cache type from file to
ext:database.

Rather than store cache data in the file system store cache data
in the database. These are the last 2 beaker caches that need
to be converted to the database as part of issue 15216.
Introduce new config parameters for database url, cache table name.
Introduce logic to default values of a parameter to values of another
parameter. i.e. default cache database url to database_connection.
Metadata for this was placed in code. Should probably be
placed eventually in config file which would require a change
to the config_schema.yml.

Introduced a new unit test, test_citations_db.py
Did not introduce a new test for the biotools_service beaker cache
because I could not find a place in the code where it was used.

claudiofr committed:
Fix testing errors after uploading changes to support a database
based beaker cache.

Change test-requirements.txt to require v beaker 1.11.0 because latest
version of beaker apparently introduced a bug that broke the ability
to use database as a cache. Also fixed lint errors.

claudiofr committed:
Fix mypy errors after uploading changes to support a database based beaker cache.

mypy kept complaining that module 'galaxy' has no attribute 'config'
on an 'from galaxy import config' statement. So I changed it to
'import galaxy.config'

claudiofr committed:
Fix lint errors after uploading changes to support a database based beaker cache.

claudiofr committed:
Fix formatting errors after uploading changes to support a database based beaker cache.

claudiofr committed:
Fix isort errors after uploading changes to support a database based beaker cache.

claudiofr committed:
Changed default beaker cache table name to beaker_cache for all caches

All 3 beaker caches, mulled_resolution, citations, biotools service,
now use the same default cache table, beaker_cache. This reduces the
number of open db connections because beaker creates a separate
sqlalchemy engine and associated connection pool for each distinct
table and each connection pool opens up a certain number of connections.
This was causing the integration tests to fail because the max
postgress connections(100) were being exceeded.

claudiofr committed:
Change default beaker cache type to file for the 3 beaker caches.

mvdbeek commented on lib/galaxy/managers/citations.py:
```diff
@@ -38,9 +38,12 @@ def _get_tool(self, tool_id):
             "cache.lock_dir": getattr(config, "citation_cache_lock_dir", None),
+            "cache.url": getattr(config, "citation_cache_url", None),
+            "cache.table_name": getattr(config, "citation_cache_table_name", None),
+            "cache.schema_name": getattr(config, "citation_cache_schema_name", None),
```
I think this should still default to file, but I also think the `getattr` isn't necessary at all and predates a refactoring of the `config` instance. All config attributes are always set now, so 

```suggestion
            "cache.type": config.citation_cache_type,
            "cache.data_dir": config.citation_cache_data_dir,
            "cache.lock_dir": config.citation_cache_lock_dir,
            "cache.url": config.citation_cache_url,
            "cache.table_name": config.citation_cache_table_name,
            "cache.schema_name": config.citation_cache_schema_name,
```
should work too

mvdbeek commented on lib/galaxy/tools/biotools.py:
```diff
@@ -16,9 +16,12 @@ def get_galaxy_biotools_metadata_source(config) -> BiotoolsMetadataSource:
         "cache.lock_dir": getattr(config, "biotools_service_cache_lock_dir", None),
+        "cache.url": getattr(config, "biotools_service_cache_url", config.database_connection),
+        "cache.table_name": getattr(config, "biotools_service_cache_table_name", None),
+        "cache.schema_name": getattr(config, "biotools_service_cache_schema_name", None),
```
```suggestion
        "cache.type": config.biotools_service_cache_type,
        "cache.data_dir": config.biotools_service_cache_data_dir,
        "cache.lock_dir": config.biotools_service_cache_lock_dir,
        "cache.url": config.biotools_service_cache_url,
        "cache.table_name": config.biotools_service_cache_table_name,
        "cache.schema_name": config.biotools_service_cache_schema_name,
```

mvdbeek commented on packages/tool_util/test-requirements.txt:
```diff
@@ -1,3 +1,4 @@
-beaker
+beaker==1.11.0 ; python_version >= "3.7" and python_version < "3.12"
+sqlalchemy==1.4.46 ; python_version >= "3.7" and python_version < "3.12"
```
The package requirements are ideally unpinned. It seems that 1.12.1 fixes your issue, so I think we should remove the pin here (ideally for sqlalchemy too, or if 2.0 fails, to pin it to `sqlalchemy<=2`)

claudiofr replied:
As far as pinning package versions in requirements.txt is concerned this has always been a best practice in my experience. In fact, the main galaxy requirements.txt specifies specific versions or ranges of versions for all referenced packages. If you do not pin a referenced package to a specific version or a version range there is no way to guarantee that what gets deployed is the same thing that was tested. This is especially true for third party packages as we have seen with the bug introduced in Beaker v1.12.0. I would think that test-requirements.txt should be consistent with the main requirements.txt file.

mvdbeek replied:
There's typically a distinction being made between libraries and applications that are shipped to users/deployers. Galaxy as an application pins all dependencies hard (i.e to specific, tested versions, see https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/dependencies/pinned-requirements.txt), while libraries should be pinned lightly, so that as the author of an application you have a good chance of being able to pick a set of compatible requirements.

And these packages here are published individually to pypi for re-use in other libraries or applications.

nsoranzo replied:
In addition to Marius' comment, the central list of compatible "light" pinnings for Galaxy Python dependencies is in `pyproject.toml` , and the various `*requirements.txt` files inside `packages/` should mirror them, see my other comment.

claudiofr committed:
Clear the beaker caches after creating them on app startup

Clear the caches in the event the cache type is database. This will
delete any rows in the cache table that could be left over from a
prior run of the app preventing an accumulation of stale data.

mvdbeek commented on lib/galaxy/app.py:
```diff
@@ -331,6 +331,8 @@ def _configure_toolbox(self):
             mulled_resolution_cache = CacheManager(**parse_cache_config_options(cache_opts)).get_cache(
                 "mulled_resolution"
             )
+            # If using database cache clear cache table contents
```
It should be persistent across processes and restarts, what was the reason to change this ?

claudiofr replied:
I guess I don't know the requirements for these caches. How long are entries expected to stay in the caches? I made this change based upon your comment in the meeting that you didn't want the default cache type to be database because you were concerned that data would accumulate in the cache table and not be cleaned out. I also assumed that in a production environment app restarts would be very infrequent. Is that not the case? So I figured this would be a good place to clean out the table to address your concern. I presume data accumulation would also be a problem with a file based cache. So why is there a concern specifically with the database cache?

mvdbeek replied:
Restarts can be frequent, for config updates, or because of resources limits, and they can be simultaneous or staggered, so I don't think this is safe.

It's not specific to the database cache, but it's safe to assume people know how to delete a file, while dropping the appropriate table requires a bit more knowledge. The lifetime is typically unlimited, but we may have to drop data occasionally, that for instance was necessary when admins upgraded python versions with incompatible pickle protocols.

mvdbeek replied:
That's a long answer to say that I would default to a file-based cache for simplicity reasons as you're doing now, and database is a good option for production deployments with kubernetes, where you can run out of (very low) file lock limits. If the cache needs to be cleared admins have to do this manually, and that's easier with files.

nsoranzo commented on packages/tool_util/test-requirements.txt:
```diff
@@ -1,3 +1,4 @@
 beaker
+sqlalchemy
```
This needs to be in sync with the pinning specified in `pyproject.toml`:

```suggestion
SQLAlchemy>=1.4.25,<2
```

claudiofr committed:
Add new config parameters for citation database beaker cache in tool_shed yaml file

claudiofr committed:
Move test_resolution_cache_db.py to test/unit/app/tools directory

The test/unit/app directory tree has a test-requirements.txt file that
brings in dependencies required by database utility modules imported by
the unit test.

nsoranzo commented on packages/tool_util/test-requirements.txt:
```diff
@@ -1,3 +1,3 @@
-beaker
+beaker==1.11.0 ; python_version >= "3.7" and python_version < "3.12"
```
```suggestion
Beaker
```

The Beaker pin in `pyproject.toml` seems unnecessary, I'll open a separate PR to verify that we can remove it from there.

nsoranzo replied:
https://github.com/galaxyproject/galaxy/pull/15665

mvdbeek committed:
Drop beaker pin

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

mvdbeek committed:
Update galaxy config schema with newer version of options

mvdbeek committed:
Rebuild schema docs
